File: android/src/main/kotlin/com/flutter_klarna_payment/flutter_klarna_payment/PaymentViewCallback.kt
package com.flutter_klarna_payment.flutter_klarna_payment

import com.klarna.mobile.sdk.api.payments.KlarnaPaymentView
import com.klarna.mobile.sdk.api.payments.KlarnaPaymentViewCallback
import com.klarna.mobile.sdk.api.payments.KlarnaPaymentsSDKError

class PaymentViewCallback(private  val streamHandler: PaymentStreamHandler) : KlarnaPaymentViewCallback {
    override fun onAuthorized(
        view: KlarnaPaymentView,
        approved: Boolean,
        authToken: String?,
        finalizedRequired: Boolean?
    ) {
        streamHandler.sendMessage("authed",null)
        println("callback debug: Authed")
        view.finalize(null)
    }

    override fun onErrorOccurred(view: KlarnaPaymentView, error: KlarnaPaymentsSDKError) {
        streamHandler.sendMessage("error",error.message)
        println("callback debug: error ${error.message}")

    }

    override fun onFinalized(view: KlarnaPaymentView, approved: Boolean, authToken: String?) {
        streamHandler.sendMessage("finalized",null)
        println("callback debug: finalized")

    }

    override fun onInitialized(view: KlarnaPaymentView) {
        streamHandler.sendMessage("initialized",null)
        streamHandler.setPaymentView(view)
        println("callback debug: initialized")

        view.load(null);
    }

    override fun onLoadPaymentReview(view: KlarnaPaymentView, showForm: Boolean) {
        streamHandler.sendMessage("loadedPaymentReview",null)
        view.loadPaymentReview();
    }

    override fun onLoaded(view: KlarnaPaymentView) {
        println("callback debug: loaded")

        streamHandler.sendMessage("loaded",null)

    }

    override fun onReauthorized(view: KlarnaPaymentView, approved: Boolean, authToken: String?) {
        println("callback debug: reauthorized")

        streamHandler.sendMessage("reauthorized",null)

    }

}

File: ios/Classes/PaymentViewCallback.swift
//
//  PaymentViewCallback.swift
//  flutter_klarna_payment
//
//  Created by qaq on 08/10/2023.
//

import Foundation

import KlarnaMobileSDK


class PaymentViewCallback: KlarnaPaymentEventListener {
    private var heightConstr: NSLayoutConstraint?
    private let streamHandler: PaymentStreamHandler

    init(streamHandler:PaymentStreamHandler){
        self.streamHandler=streamHandler
    }
    
    func setHeightConstr(heightConstr:NSLayoutConstraint){
        self.heightConstr = heightConstr
    }
    
    func klarnaResized(paymentView: KlarnaMobileSDK.KlarnaPaymentView, to newHeight: CGFloat) {
        
        heightConstr?.constant=newHeight
        
    }
    
    
    
    func klarnaInitialized(paymentView: KlarnaMobileSDK.KlarnaPaymentView) {
        streamHandler.sendMessage(state: "initialized", message: nil);
         paymentView.load(jsonData: nil)
        streamHandler.setPaymentView(paymentView)
     }
     
     func klarnaLoaded(paymentView: KlarnaMobileSDK.KlarnaPaymentView) {
         streamHandler.sendMessage(state:"loaded", message: nil as String?);
     }
     
     func klarnaLoadedPaymentReview(paymentView: KlarnaMobileSDK.KlarnaPaymentView) {
         
         
         streamHandler.sendMessage(state:"paymentReview", message: nil);
     }
     
     func klarnaAuthorized(paymentView: KlarnaMobileSDK.KlarnaPaymentView, approved: Bool, authToken: String?, finalizeRequired: Bool) {
         if (authToken != nil) {
             streamHandler.sendMessage(state:"authorized", message: authToken);

         }
     }
     
     func klarnaReauthorized(paymentView: KlarnaMobileSDK.KlarnaPaymentView, approved: Bool, authToken: String?) {
         streamHandler.sendMessage(state:"reauthorized", message: nil);

     }
     
     func klarnaFinalized(paymentView: KlarnaMobileSDK.KlarnaPaymentView, approved: Bool, authToken: String?) {
         streamHandler.sendMessage(state:"finalized", message: nil);

     }
     
     func klarnaFailed(inPaymentView paymentView: KlarnaMobileSDK.KlarnaPaymentView, withError error: KlarnaMobileSDK.KlarnaPaymentError) {
         streamHandler.sendMessage(state:"failed", message: error.message);

         
     }
    
}
